bip-0178.mediawiki

```diff
@@ -0,0 +1,79 @@
+<pre>
+  BIP: 178
+  Layer: Applications
+  Title: Typed Private Keys
+  Author: Karl-Johan Alm <karljohan-alm@garage.co.jp>
+  Comments-Summary: No comments yet.
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0178
+  Status: Draft
+  Type: Standards Track
+  Created: 2018-04-04
+  License: CC0-1.0
+</pre>
+
+== Abstract ==
+
+An extension to the private key (WIF) format to specify what kind of public key the private key corresponds to.
+
+== Motivation ==
+
+There are several types of public keys which can all be associated with a given private key: P2PKH (legacy <code>1...</code> format), P2SH-P2WPKH (SegWit public key inside P2SH), P2WPKH (bech32), etc.
+
+While private keys have a 1-byte suffix indicating whether the corresponding public key is compressed (<code>0x01</code>) or not (through suffix absence), there is no way of knowing what kind of public keys were associated with the private key. As a result, when importing a private key, the wallet has to assume all kinds, and keep track of each possible alternative.
+
+By extending the suffix, we can specify what kind of public key was associated with the given private key.
+
+== Specification ==
+
+Currently, private keys are stored as a uint256 (private key data) followed by an optional uint8 (compressed flag). The latter is extended to specify the public key types:
+
+{|class="wikitable" style="text-align: center;"
+|-
+!Value
+!Type
+!Compr
+!Clarification
+|-
+|No suffix||P2PKH_UNCOMPRESSED||No||Uncompressed legacy public key. Unknown public key format
+|-
+|<code>0x01</code>||P2PKH_COMPRESSED||Yes||Compressed legacy public key. Unknown public key format
+|-
+|<code>0x10</code>||P2PKH||Yes||Compressed legacy public key. Legacy public key format (<code>1...</code>)
+|-
+|<code>0x11</code>||P2WPKH||Yes||Bech32 format (native Segwit)
+|-
+|<code>0x12</code>||P2WPKH_P2SH||Yes||Segwit nested in BIP16 P2SH (<code>3...</code>)
+|-
+|<code>0x13</code>||P2SH|| &mdash; ||Non-Segwit BIP16 P2SH (<code>3...</code>)
```

Comment:
What script type does this refer to? The slip32 prefixes don't allow arbitrary redeem scripts (it would defeat the point IMO). If this is meant to be P2PKH_P2SH, please refer to it as that!

Reply:
@afk11 IMHO it should be P2SH_P2PKH,  as better represents P2SH(P2PKH)

Reply:
I believe this would be used for arbitrary P2SH scripts with a related private key (e.g. multisig). Is there a reason to even have a P2PKH in P2SH variant?

Reply:
@kallewoof given the lack of redeem script context, wouldn't this simply be a matter of unsupported by any implementation?

Otherwise we are assuming the context and it is no better than providing a standard WIF?

Reply:
Well, if I make a custom multisig with someone else, and I export the private key, even if I don't know the redeem script itself, it may be useful to know what the key was used for. An implementation that doesn't support custom redeem scripts would probably keep it unimplemented, but that's a good thing, right? They can't spend the UTXOs anyway.

Reply:
@kallewoof but additionally they have no assurance that the key is actually using _their_ redeem scripts either.
Its a back-to-square one issue,  an implementation could _always_ check every type to see if their are transactions on those addresses,  but this was about marking the WIF with a 'guarantee' to avoid that, no?

Reply:
Yes, that is true. For completeness sake, there is something to be gained from having every type listed, even if you don't have all the details. Contrarily, having a custom script with an associated private key, which type would it be, if we removed the P2SH type?

Reply:
@kallewoof I think there is no point in having a flag to say "this is some form of P2SH but you don't know what". If you're going to need to give additional information anyway, you might as well just give the private key in WIF format plus whatever other information is needed.

My view is that this typed WIF format is really just a shorthand notation for "here is a private key, and I've conveniently packed the address alongside it". If you can't use the shorthand, don't.

Reply:
So basically, use 0x01 for keys that are in custom stuff? That means wallets will import all types rather than import nothing, knowing they will never see the UTXO.

Reply:
Yes, my point is that if you're using a custom script/address, you *must* have more information. There is simply no way you could hope to have everything in one string. And there is no need for that, I think.

In the general case, you will provide your wallet with a bunch of information to describe what outputs you mean, and which private keys are relevant for it. In a few exceptionally simple (but remarkably common) cases, you get to put everything in one string.

Reply:
So, conclusively, do not include types that we won't be able to use without additional information anyway. I.e. remove P2SH, P2WSH, and P2WSH_P2SH, and keep the rest.

Reply:
And then use 0x01=P2PKH_COMPRESSED for 'custom stuff' exactly like is done today.

Reply:
Updated BIP, removing the 3 script-dependent types.

Reply:
@kallewoof I think having script-hash types is fine once the flag tells you what the redeemscript is. I see no harm in having P2SH_P2WPKH, P2WSH_P2PKH, and if we support producing those scriptPubKeys. Script hash scripts have no degrees of freedom besides the script they commit to, so IMO supporting script-hash extensions is pretty natural.

We could leave it the way it was before, assuming they refer to P2PKH.. 
Or we could use bits 2 and 3 to indicate P2SH and P2WSH, and maybe use others to indicate the scriptPubKey type?

Another, is just ignore script-hash usage entirely, and assume the software will deal with it.